### DISCUSSION TITLE
Newbie question - Request for simple julia data loader

### DISCUSSION THREAD

**Discussion by gganapat** [observablehq/framework#1062] [Q&A] [answered]
A sample julia code data loader would be much appreciated. For example printing a text "Hello world" I had a file with
```
print("Hello world")
```
and named it "test.csv.jl" in /docs/data
* If the output is text is it still ok to call the data loader test.csv.jl? 
* And is the #!/usr/bin/env julia header needed to get julia to execute the command? 
* And I assume this will work only if npm run build is invoked? 

I was expecting a file to be generated in /docs/.observablehq/cache but nothing is there when I run npm run build. Any help would be appreciated.

Thanks!

**Comment by mbostock** [comment, marked as answer]
Yes, you can have a data loader named `test.csv.jl` and not have it generate CSV — it’s just a little confusing. If you just want a test, maybe call it `test.txt.jl` instead?

No, the `#!/usr/bin/env julia` shebang is not necessary if you use the `.jl` extension for your data loader. A shebang is only necessary when using an executable (`.exe`) data loader rather than an interpreted one (and in that case, your data loader must also have the executable bit set).

Data loaders are only run if you reference their output. So in addition to writing the data loader, you’ll need to load the file in a Markdown page (or in a JavaScript module that is imported by a Markdown page). If you visit that page in your preview server, it will run your data loader and cache the output. Likewise if you build the site (`npm run build`), it will run your data loader.

So, if this is your data loader, `hello-julia.txt.jl`:

```jl
print("Hello world")
```

Here is a corresponding Markdown page that invokes it:

````md
# Hello, Julia

```js echo
FileAttachment("hello-julia.txt").text()
```
````

You should then see this:

<img width="1228" alt="Screenshot 2024-03-14 at 10 24 14 PM" src="https://github.com/observablehq/framework/assets/230541/b0547247-ad06-4063-9b46-5bb79cccdc0c">

**Comment by gganapat** [comment]
Thank you!!

**Comment by gganapat** [comment]
I'm sorry I don't know what I'm doing wrong. I did exactly what you wrote, including just copy/pasting your data-loader file hello-julia.txt.jl into /docs/data folder, created same julia.md file as you did, and ran ```npm run dev``` and this is what I get

<img width="965" alt="Screenshot 2024-03-15 at 11 30 30 AM" src="https://github.com/observablehq/framework/assets/46061218/2ac3f3aa-2eb3-4a26-adf0-d246a1e4f804">

Running ```npm run build``` and followed by ```npm run dev``` doesn't help either. There is something about the architecture I'm not able to get a grasp on.

Checking to see if it was a problem with my julia installation, I got no errors.

<img width="577" alt="Screenshot 2024-03-15 at 11 40 34 AM" src="https://github.com/observablehq/framework/assets/46061218/09726fe0-950f-4752-80d3-7f45505d4e9e">


Thanks,
Gani --

**Reply by mbostock** [reply]
The path you give to `FileAttachment` needs to match the relative path from the Markdown file to the data loader. So, if your Markdown file is in `docs/julia.md`, and your data loader is in `docs/data/hello-julia.txt.jl`, then you need to say

```js
FileAttachment("data/hello-julia.txt").text()
```

You can only say `FileAttachment("hello-julia.txt")` if your data loader is at `docs/hello-julia.txt.jl`.

**Reply by gganapat** [reply]
I did put in docs folder first, then did with /docs/data. But both didn't work.
<img width="995" alt="Screenshot 2024-03-15 at 11 59 38 AM" src="https://github.com/observablehq/framework/assets/46061218/4b246bbf-11d8-4f4d-8c47-4f7256fccd1b">

It appears to hang on a promise as seen in this output---

<img width="745" alt="Screenshot 2024-03-15 at 12 00 18 PM" src="https://github.com/observablehq/framework/assets/46061218/88c3b86c-66e3-4995-8a21-39c43bf21b8c">


which I confirmed when I changed julia.md file to 

<img width="488" alt="Screenshot 2024-03-15 at 12 02 58 PM" src="https://github.com/observablehq/framework/assets/46061218/042323af-424e-4003-86cd-3484c4f4acea">

And the julia.md page as below ---

<img width="1015" alt="Screenshot 2024-03-15 at 12 03 37 PM" src="https://github.com/observablehq/framework/assets/46061218/6b21395f-e1f3-4f24-8187-a220a8c0d4f4">

**Reply by gganapat** [reply]
And with ```const text = FileAttachment("data/hello-julia.txt").text();``` (no / before data) same result.

**Reply by mbostock** [reply]
You’ve added a leading slash and a trailing semicolon. Here is the code again (with no semicolon so you get implicit display):

```js
FileAttachment("data/hello-julia.txt").text()
```

Alternatively you can make the `await` and `display` explicit as:

```js
const text = await FileAttachment("data/hello-julia.txt").text();

display(text);
```

**Reply by gganapat** [reply]
ok, now I'm convinced I'm cursed :)

Data loader

<img width="518" alt="Screenshot 2024-03-15 at 1 17 42 PM" src="https://github.com/observablehq/framework/assets/46061218/f522fa3b-2873-410f-a42f-ac4811c1a32d">



Path for data loader:

<img width="578" alt="Screenshot 2024-03-15 at 1 17 32 PM" src="https://github.com/observablehq/framework/assets/46061218/e998ca26-2863-414b-b0b2-673abaa514ee">


julia.md

<img width="950" alt="Screenshot 2024-03-15 at 1 17 57 PM" src="https://github.com/observablehq/framework/assets/46061218/286f55c5-0b96-4d78-8be9-106183a52245">

From terminal

<img width="342" alt="Screenshot 2024-03-15 at 1 24 06 PM" src="https://github.com/observablehq/framework/assets/46061218/590f2b04-12ae-4324-949c-070f0eee210f">

**Reply by mbostock** [reply]
Are you using an old version of Framework (pre-1.3.0) that doesn’t have built-in support for Julia data loaders maybe?

**Comment by gganapat** [comment]
I'll download latest one and try again. Thanks for your patience.
gani -

**Comment by gganapat** [comment]
ok, it works now! I was probably using the older version without support. Would the results have been the same if I had the data loader hello-julia.txt.sh? 
```
#! /usr/bin/env julia
print("Hello from Julia");
```
in the older version? I seem to remember that being the case.
Thanks again!
gani -

**Reply by mbostock** [reply]
For a shebang you need to use an executable (`.exe`) data loader. A shell (`.sh`) data loader has to be written in your shell’s language (such as bash). An executable data loader also requires the executable bit to be set. So you could put that in `hello-julia.txt.exe` and then say `chmod +x hello-julia.txt.exe` and it would work.

**Reply by gganapat** [reply]
I'm on a Mac, so would it still be ```.exe```? I would have thought it should be ```hello-julia.txt.sh``` with right permission as above.

**Reply by gganapat** [reply]
I confirm it works with .exe. Thanks!
